{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "be43dbb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import gffutils\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "\n",
    "in_dir = \"input\"\n",
    "out_dir = \"output\"\n",
    "Path(out_dir).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "940b2152",
   "metadata": {},
   "source": [
    "## Create GFF databases and read in features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2ca85022",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_db(path, db_filename, **kwargs):\n",
    "    \"\"\"\n",
    "    Create/connect db for given gff file path\n",
    "    \"\"\"\n",
    "    if not os.path.isfile(os.path.join(out_dir, db_filename)):\n",
    "        db = gffutils.create_db(path, os.path.join(out_dir, db_filename), force=True, **kwargs)\n",
    "    else:\n",
    "        db = sqlite3.connect(os.path.join(out_dir, db_filename), check_same_thread=False)\n",
    "        db = gffutils.FeatureDB(db)\n",
    "    return db\n",
    "\n",
    "# WormBase release WS283 canonical three_prime_UTR for C. elegans chromosome I\n",
    "control_fn = os.path.join(in_dir, \"WormBase.I.three_prime_UTR.gff3\")\n",
    "db_control = create_db(control_fn, \"control.db\")\n",
    "control = list(db_control.all_features())\n",
    "\n",
    "# Output from GETUTR run with PAVA smoothing and s >= 0.95\n",
    "test_fn = os.path.join(in_dir, \"getutr_output.gff3\")\n",
    "db_test = create_db(test_fn, \"test_getutr.db\")\n",
    "test = list(db_test.all_features())\n",
    "\n",
    "# WormBase release WS283 canonical mRNA transcripts for C. elegans chromosome I\n",
    "genes_fn = os.path.join(in_dir, \"WormBase.I.mRNA.gff3\")\n",
    "db_genes = create_db(genes_fn, \"genes.db\")\n",
    "genes = list(db_genes.all_features())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6119982",
   "metadata": {},
   "source": [
    "## Prepare control dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6fb1344",
   "metadata": {},
   "source": [
    "C. elegans contains alternative mRNA transcripts for each gene, and the control UTRs refer only to these mRNA transcript \"names\" in their attributes. GETUTR UTRs are represented by fragmented \"sequence_features\", each with an associated \"s\" score and referring only to these mRNA transcript \"names\" in their attributes. Therefore, it will be helpful to create a sequence name -> gene id mapping dataframe to allow comparisions between the two datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f2cd9eca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y74C9A.3.1</td>\n",
       "      <td>WBGene00022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y74C9A.2a.3</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y74C9A.2a.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y74C9A.2a.2</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y74C9A.2b.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4698</th>\n",
       "      <td>F31C3.1.1</td>\n",
       "      <td>WBGene00000881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4699</th>\n",
       "      <td>F31C3.3.1</td>\n",
       "      <td>WBGene00009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4700</th>\n",
       "      <td>F31C3.4.1</td>\n",
       "      <td>WBGene00009286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4701</th>\n",
       "      <td>F31C3.5.1</td>\n",
       "      <td>WBGene00009287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4702</th>\n",
       "      <td>F31C3.6a.1</td>\n",
       "      <td>WBGene00009288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4703 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             name            gene\n",
       "0      Y74C9A.3.1  WBGene00022277\n",
       "1     Y74C9A.2a.3  WBGene00022276\n",
       "2     Y74C9A.2a.1  WBGene00022276\n",
       "3     Y74C9A.2a.2  WBGene00022276\n",
       "4     Y74C9A.2b.1  WBGene00022276\n",
       "...           ...             ...\n",
       "4698    F31C3.1.1  WBGene00000881\n",
       "4699    F31C3.3.1  WBGene00009285\n",
       "4700    F31C3.4.1  WBGene00009286\n",
       "4701    F31C3.5.1  WBGene00009287\n",
       "4702   F31C3.6a.1  WBGene00009288\n",
       "\n",
       "[4703 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf = pd.DataFrame([vars(c) for c in genes], columns=[\"featuretype\", \"start\", \"end\", \"strand\", \"attributes\"])\n",
    "gdf = gdf[gdf[\"featuretype\"] == \"mRNA\"]\n",
    "gdf[\"gene\"] = gdf[\"attributes\"].apply(lambda x: x.get(\"Parent\")[0].split(\":\")[-1])\n",
    "gdf[\"name\"] = gdf[\"attributes\"].apply(lambda x: x.get(\"Name\")[0])\n",
    "gdf = gdf[[\"name\", \"gene\"]]\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ca3ec53",
   "metadata": {},
   "source": [
    "Create a dataframe of canonical WormBase control UTRs with a \"name\" column consisting of the parent gene's name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2e3b122f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4116</td>\n",
       "      <td>4220</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.3.1</td>\n",
       "      <td>WBGene00022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16586</td>\n",
       "      <td>16833</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.2</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16586</td>\n",
       "      <td>16837</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16702</td>\n",
       "      <td>16793</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.3</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17484</td>\n",
       "      <td>17910</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.4a.1</td>\n",
       "      <td>WBGene00022278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4192</th>\n",
       "      <td>15040474</td>\n",
       "      <td>15040608</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.1.1</td>\n",
       "      <td>WBGene00000881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>15041433</td>\n",
       "      <td>15041884</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.3.1</td>\n",
       "      <td>WBGene00009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>15050797</td>\n",
       "      <td>15051145</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.4.1</td>\n",
       "      <td>WBGene00009286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4195</th>\n",
       "      <td>15053608</td>\n",
       "      <td>15053793</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.6a.1</td>\n",
       "      <td>WBGene00009288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4196</th>\n",
       "      <td>15053619</td>\n",
       "      <td>15053672</td>\n",
       "      <td>+</td>\n",
       "      <td>F31C3.5.1</td>\n",
       "      <td>WBGene00009287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4197 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         start       end strand         name            gene\n",
       "0         4116      4220      -   Y74C9A.3.1  WBGene00022277\n",
       "1        16586     16833      +  Y74C9A.2a.2  WBGene00022276\n",
       "2        16586     16837      +  Y74C9A.2a.1  WBGene00022276\n",
       "3        16702     16793      +  Y74C9A.2a.3  WBGene00022276\n",
       "4        17484     17910      -  Y74C9A.4a.1  WBGene00022278\n",
       "...        ...       ...    ...          ...             ...\n",
       "4192  15040474  15040608      -    F31C3.1.1  WBGene00000881\n",
       "4193  15041433  15041884      -    F31C3.3.1  WBGene00009285\n",
       "4194  15050797  15051145      -    F31C3.4.1  WBGene00009286\n",
       "4195  15053608  15053793      -   F31C3.6a.1  WBGene00009288\n",
       "4196  15053619  15053672      +    F31C3.5.1  WBGene00009287\n",
       "\n",
       "[4197 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdf = pd.DataFrame([vars(c) for c in control], columns=[\"start\", \"end\", \"source\", \"strand\", \"attributes\"])\n",
    "cdf = cdf[cdf.source == \"WormBase\"]\n",
    "cdf['name'] = cdf[\"attributes\"].apply(lambda x: x.get(\"Parent\")[0].split(\":\")[-1])\n",
    "cdf = cdf.drop(['attributes', \"source\"], axis=1)\n",
    "cdf = pd.merge(cdf, gdf, on=\"name\")\n",
    "cdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fe678e2",
   "metadata": {},
   "source": [
    "Aggregate transcripts for minimum start base and maximum end base to capture full extent of the UTR for each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fa6e5c2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000001</th>\n",
       "      <td>+</td>\n",
       "      <td>5109948</td>\n",
       "      <td>5110183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000006</th>\n",
       "      <td>-</td>\n",
       "      <td>2571740</td>\n",
       "      <td>2571987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000010</th>\n",
       "      <td>-</td>\n",
       "      <td>11388682</td>\n",
       "      <td>11389140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000012</th>\n",
       "      <td>-</td>\n",
       "      <td>3886070</td>\n",
       "      <td>3886110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000020</th>\n",
       "      <td>-</td>\n",
       "      <td>2255921</td>\n",
       "      <td>2256151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00304996</th>\n",
       "      <td>+</td>\n",
       "      <td>5341995</td>\n",
       "      <td>5342045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00305018</th>\n",
       "      <td>-</td>\n",
       "      <td>9588662</td>\n",
       "      <td>9588719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306080</th>\n",
       "      <td>-</td>\n",
       "      <td>2939477</td>\n",
       "      <td>2939518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306081</th>\n",
       "      <td>+</td>\n",
       "      <td>6872119</td>\n",
       "      <td>6872142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306121</th>\n",
       "      <td>-</td>\n",
       "      <td>231578</td>\n",
       "      <td>231691</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2593 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end\n",
       "gene                                     \n",
       "WBGene00000001      +   5109948   5110183\n",
       "WBGene00000006      -   2571740   2571987\n",
       "WBGene00000010      -  11388682  11389140\n",
       "WBGene00000012      -   3886070   3886110\n",
       "WBGene00000020      -   2255921   2256151\n",
       "...               ...       ...       ...\n",
       "WBGene00304996      +   5341995   5342045\n",
       "WBGene00305018      -   9588662   9588719\n",
       "WBGene00306080      -   2939477   2939518\n",
       "WBGene00306081      +   6872119   6872142\n",
       "WBGene00306121      -    231578    231691\n",
       "\n",
       "[2593 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_utrs = cdf.groupby([\"gene\", \"strand\"]).agg({\"start\": min, \"end\": max}).reset_index(\"strand\")\n",
    "control_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a2bf685",
   "metadata": {},
   "source": [
    "## Prepare test dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a67a40dd",
   "metadata": {},
   "source": [
    "Create a dataframe of GETUTR test UTRs with a \"gene\" column consisting of the parent gene's name. We will aggregate the \"sequence_features\" for each gene name for minimum start base and maximum end base to capture the full extent of the UTR for that gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d80996d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000001</th>\n",
       "      <td>+</td>\n",
       "      <td>5109949</td>\n",
       "      <td>5110316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000023</th>\n",
       "      <td>+</td>\n",
       "      <td>12018641</td>\n",
       "      <td>12018859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000079</th>\n",
       "      <td>+</td>\n",
       "      <td>7774293</td>\n",
       "      <td>7777525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000084</th>\n",
       "      <td>+</td>\n",
       "      <td>7607752</td>\n",
       "      <td>7609660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000096</th>\n",
       "      <td>+</td>\n",
       "      <td>9531100</td>\n",
       "      <td>9533701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303032</th>\n",
       "      <td>-</td>\n",
       "      <td>4647835</td>\n",
       "      <td>4650729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303056</th>\n",
       "      <td>-</td>\n",
       "      <td>7706598</td>\n",
       "      <td>7706960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303057</th>\n",
       "      <td>-</td>\n",
       "      <td>5862563</td>\n",
       "      <td>5864973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303058</th>\n",
       "      <td>-</td>\n",
       "      <td>14049858</td>\n",
       "      <td>14050372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303078</th>\n",
       "      <td>-</td>\n",
       "      <td>9074692</td>\n",
       "      <td>9074869</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2176 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end\n",
       "gene                                     \n",
       "WBGene00000001      +   5109949   5110316\n",
       "WBGene00000023      +  12018641  12018859\n",
       "WBGene00000079      +   7774293   7777525\n",
       "WBGene00000084      +   7607752   7609660\n",
       "WBGene00000096      +   9531100   9533701\n",
       "...               ...       ...       ...\n",
       "WBGene00303032      -   4647835   4650729\n",
       "WBGene00303056      -   7706598   7706960\n",
       "WBGene00303057      -   5862563   5864973\n",
       "WBGene00303058      -  14049858  14050372\n",
       "WBGene00303078      -   9074692   9074869\n",
       "\n",
       "[2176 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdf = pd.DataFrame([vars(t) for t in test], columns=[\"start\", \"end\", \"strand\", \"attributes\"])\n",
    "tdf['name'] = tdf[\"attributes\"].apply(lambda x: x.get(\"name\")[0])\n",
    "tdf = tdf.drop(['attributes'], axis=1)\n",
    "tdf = pd.merge(tdf, gdf, on=\"name\")\n",
    "test_utrs = tdf.groupby([\"strand\", \"gene\"]).agg({\"start\": min, \"end\": max}).reset_index(\"strand\")\n",
    "test_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc4d603f",
   "metadata": {},
   "source": [
    "Write out a GETUTR output with aggregated UTRs. This is useful for visualisation in Artemis, as well as being necessary to call GFFCompare."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d4432b43",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(out_dir, \"getutr_output.fixed.gff3\"), \"w\") as f:\n",
    "    for idx, (gid, row) in enumerate(test_utrs.iterrows(), 1):\n",
    "        feature = gffutils.Feature(\n",
    "            seqid=\"I\",\n",
    "            source=\"GETUTR\",\n",
    "            featuretype=\"three_prime_UTR\",\n",
    "            start=row.start,\n",
    "            end=row.end,\n",
    "            score=\".\",\n",
    "            strand=row.strand,\n",
    "            frame=\".\",\n",
    "            attributes={\"Name\": [gid], \"Parent\": [\"Gene:\" + gid], \"ID\": [\"three_prime_UTR_{}\".format(idx)] }\n",
    "        )\n",
    "        f.write(str(feature) + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9447c55f",
   "metadata": {},
   "source": [
    "# Compare the two datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c9fdcfe",
   "metadata": {},
   "source": [
    "Now that we have two comparable datasets, we can calculate some statistical differences. We determine length of each UTR in both dataframes, and determine differences of test UTR length to true UTR length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "592d8b04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>len</th>\n",
       "      <th>difference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000001</th>\n",
       "      <td>+</td>\n",
       "      <td>5109949</td>\n",
       "      <td>5110316</td>\n",
       "      <td>367</td>\n",
       "      <td>132.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000023</th>\n",
       "      <td>+</td>\n",
       "      <td>12018641</td>\n",
       "      <td>12018859</td>\n",
       "      <td>218</td>\n",
       "      <td>174.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000079</th>\n",
       "      <td>+</td>\n",
       "      <td>7774293</td>\n",
       "      <td>7777525</td>\n",
       "      <td>3232</td>\n",
       "      <td>3047.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000084</th>\n",
       "      <td>+</td>\n",
       "      <td>7607752</td>\n",
       "      <td>7609660</td>\n",
       "      <td>1908</td>\n",
       "      <td>1414.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000096</th>\n",
       "      <td>+</td>\n",
       "      <td>9531100</td>\n",
       "      <td>9533701</td>\n",
       "      <td>2601</td>\n",
       "      <td>2568.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303032</th>\n",
       "      <td>-</td>\n",
       "      <td>4647835</td>\n",
       "      <td>4650729</td>\n",
       "      <td>2894</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303056</th>\n",
       "      <td>-</td>\n",
       "      <td>7706598</td>\n",
       "      <td>7706960</td>\n",
       "      <td>362</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303057</th>\n",
       "      <td>-</td>\n",
       "      <td>5862563</td>\n",
       "      <td>5864973</td>\n",
       "      <td>2410</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303058</th>\n",
       "      <td>-</td>\n",
       "      <td>14049858</td>\n",
       "      <td>14050372</td>\n",
       "      <td>514</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00303078</th>\n",
       "      <td>-</td>\n",
       "      <td>9074692</td>\n",
       "      <td>9074869</td>\n",
       "      <td>177</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2176 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end   len  difference\n",
       "gene                                                       \n",
       "WBGene00000001      +   5109949   5110316   367       132.0\n",
       "WBGene00000023      +  12018641  12018859   218       174.0\n",
       "WBGene00000079      +   7774293   7777525  3232      3047.0\n",
       "WBGene00000084      +   7607752   7609660  1908      1414.0\n",
       "WBGene00000096      +   9531100   9533701  2601      2568.0\n",
       "...               ...       ...       ...   ...         ...\n",
       "WBGene00303032      -   4647835   4650729  2894         NaN\n",
       "WBGene00303056      -   7706598   7706960   362         NaN\n",
       "WBGene00303057      -   5862563   5864973  2410         NaN\n",
       "WBGene00303058      -  14049858  14050372   514         NaN\n",
       "WBGene00303078      -   9074692   9074869   177         NaN\n",
       "\n",
       "[2176 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_utrs[\"len\"] = abs(control_utrs[\"end\"] - control_utrs[\"start\"])\n",
    "test_utrs[\"len\"] = abs(test_utrs[\"end\"] - test_utrs[\"start\"])\n",
    "test_utrs[\"difference\"] = (test_utrs[\"len\"] - control_utrs[\"len\"])\n",
    "test_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ccd2e2f",
   "metadata": {},
   "source": [
    "## Supplementary Table 2 statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa4a4051",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical utrs that are missed by GETUTR: 568 (21.91%)\n",
      "GETUTR utrs that match true utr to within 50 bases: 155 (5.98%)\n",
      "GETUTR utrs that reduce true utr by >50 bases: 513 (19.78%)\n",
      "GETUTR utrs that extend true utr by between 50 and 200 bases: 133 (5.13%)\n",
      "GETUTR utrs that extend true utr by >200 bases: 1224 (47.20%)\n",
      "GETUTR utrs called where no true utr existed previously: 151\n",
      "Total GETUTR UTRs: 2176\n"
     ]
    }
   ],
   "source": [
    "total_test = len(test_utrs)\n",
    "total_control = len(control_utrs)\n",
    "\n",
    "genes_missed_utr = control_utrs.index.difference(test_utrs.index)\n",
    "num_missed_utrs = len(genes_missed_utr)\n",
    "genes_no_utr = set(gdf[\"gene\"]).difference(control_utrs.index)\n",
    "novel_utr_genes = genes_no_utr.intersection(test_utrs.index)\n",
    "num_novel_utrs = len(novel_utr_genes)\n",
    "\n",
    "BP1 = 50\n",
    "BP2 = 200\n",
    "genes_matched_utr =  test_utrs[abs(test_utrs[\"difference\"]) <= BP1].index\n",
    "num_matched = len(genes_matched_utr)\n",
    "num_reduced = len(test_utrs[test_utrs['difference'] < -BP1])\n",
    "num_extended1 = len(test_utrs[(test_utrs['difference'] > BP1) & (test_utrs['difference'] <= BP2)])\n",
    "num_extended2 = len(test_utrs[test_utrs['difference'] > BP2])\n",
    "\n",
    "\n",
    "print(\"Canonical utrs that are missed by GETUTR: {} ({:.2f}%)\".format(num_missed_utrs, 100*num_missed_utrs/total_control))\n",
    "print(\"GETUTR utrs that match true utr to within {} bases: {} ({:.2f}%)\".format(BP1, num_matched, 100*num_matched/total_control))\n",
    "print(\"GETUTR utrs that reduce true utr by >{} bases: {} ({:.2f}%)\".format(BP1, num_reduced, 100*num_reduced/total_control))\n",
    "print(\"GETUTR utrs that extend true utr by between {} and {} bases: {} ({:.2f}%)\".format(BP1, BP2, num_extended1, 100*num_extended1/total_control))\n",
    "print(\"GETUTR utrs that extend true utr by >{} bases: {} ({:.2f}%)\".format(BP2, num_extended2, 100*num_extended2/total_control))\n",
    "print(\"GETUTR utrs called where no true utr existed previously: {}\".format(num_novel_utrs))\n",
    "print(\"Total GETUTR UTRs: {}\".format(total_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "408e532b",
   "metadata": {},
   "source": [
    "Save lists of gene ids with novel, matched and missed 3' UTRs called by GETUTR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cddb9383",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(out_dir, \"novel_utr_genes_getutr.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(novel_utr_genes)):\n",
    "        f.write(g + \"\\n\")\n",
    "\n",
    "with open(os.path.join(out_dir, \"matched_utr_genes_getutr.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(genes_matched_utr)):\n",
    "        f.write(g + \"\\n\")\n",
    "\n",
    "with open(os.path.join(out_dir, \"missed_utr_genes_getutr.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(genes_missed_utr)):\n",
    "        f.write(g + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
