{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "458728fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import gffutils\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "\n",
    "in_dir = \"input\"\n",
    "out_dir = \"output\"\n",
    "Path(out_dir).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "353857df",
   "metadata": {},
   "source": [
    "## Create GFF databases and read in features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "59ae4ad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_db(path, db_filename, **kwargs):\n",
    "    \"\"\"\n",
    "    Create/connect db for given gff file path\n",
    "    \"\"\"\n",
    "    if not os.path.isfile(os.path.join(out_dir, db_filename)):\n",
    "        db = gffutils.create_db(path, os.path.join(out_dir, db_filename), force=True, **kwargs)\n",
    "    else:\n",
    "        db = sqlite3.connect(os.path.join(out_dir, db_filename), check_same_thread=False)\n",
    "        db = gffutils.FeatureDB(db)\n",
    "    return db\n",
    "\n",
    "# WormBase release WS283 canonical three_prime_UTR for C. elegans chromosome I\n",
    "control_fn = os.path.join(in_dir, \"WormBase.I.three_prime_UTR.gff3\")\n",
    "db_control = create_db(control_fn, \"control.db\")\n",
    "control = list(db_control.all_features())\n",
    "\n",
    "# Output from UTRme run\n",
    "test = os.path.join(in_dir, \"utrme_output.gff3\")\n",
    "db_test = create_db(test, \"test_utrme.db\")\n",
    "test = list(db_test.all_features())\n",
    "\n",
    "# WormBase release WS283 canonical mRNA transcripts for C. elegans chromosome I\n",
    "genes_fn = os.path.join(in_dir, \"WormBase.I.mRNA.gff3\")\n",
    "db_genes = create_db(genes_fn, \"genes.db\")\n",
    "genes = list(db_genes.all_features())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dea57927",
   "metadata": {},
   "source": [
    "## Prepare test dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca342301",
   "metadata": {},
   "source": [
    "C. elegans contains alternative mRNA transcripts for each gene, and the UTRme UTRs refer only to these mRNA transcript \"names\" in their attributes. Therefore, it will be helpful to create a sequence name -> gene id mapping dataframe to allow comparisions between the two datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bfd56f5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y74C9A.3.1</td>\n",
       "      <td>WBGene00022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y74C9A.2a.3</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y74C9A.2a.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y74C9A.2a.2</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y74C9A.2b.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4698</th>\n",
       "      <td>F31C3.1.1</td>\n",
       "      <td>WBGene00000881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4699</th>\n",
       "      <td>F31C3.3.1</td>\n",
       "      <td>WBGene00009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4700</th>\n",
       "      <td>F31C3.4.1</td>\n",
       "      <td>WBGene00009286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4701</th>\n",
       "      <td>F31C3.5.1</td>\n",
       "      <td>WBGene00009287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4702</th>\n",
       "      <td>F31C3.6a.1</td>\n",
       "      <td>WBGene00009288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4703 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             name            gene\n",
       "0      Y74C9A.3.1  WBGene00022277\n",
       "1     Y74C9A.2a.3  WBGene00022276\n",
       "2     Y74C9A.2a.1  WBGene00022276\n",
       "3     Y74C9A.2a.2  WBGene00022276\n",
       "4     Y74C9A.2b.1  WBGene00022276\n",
       "...           ...             ...\n",
       "4698    F31C3.1.1  WBGene00000881\n",
       "4699    F31C3.3.1  WBGene00009285\n",
       "4700    F31C3.4.1  WBGene00009286\n",
       "4701    F31C3.5.1  WBGene00009287\n",
       "4702   F31C3.6a.1  WBGene00009288\n",
       "\n",
       "[4703 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf = pd.DataFrame([vars(c) for c in genes], columns=[\"featuretype\", \"start\", \"end\", \"strand\", \"attributes\"])\n",
    "gdf = gdf[gdf[\"featuretype\"] == \"mRNA\"]\n",
    "gdf[\"gene\"] = gdf[\"attributes\"].apply(lambda x: x.get(\"Parent\")[0].split(\":\")[-1])\n",
    "gdf[\"name\"] = gdf[\"attributes\"].apply(lambda x: x.get(\"Name\")[0])\n",
    "gdf = gdf[[\"name\", \"gene\"]]\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb7f3580",
   "metadata": {},
   "source": [
    "Create a dataframe of UTRme test UTRs with a \"gene\" column consisting of the parent gene's id."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3ed2df20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22939</td>\n",
       "      <td>24650</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.4a.1</td>\n",
       "      <td>WBGene00022278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>80344</td>\n",
       "      <td>81000</td>\n",
       "      <td>+</td>\n",
       "      <td>Y48G1C.2b.1</td>\n",
       "      <td>WBGene00000812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>101245</td>\n",
       "      <td>101504</td>\n",
       "      <td>-</td>\n",
       "      <td>Y48G1C.8a.1</td>\n",
       "      <td>WBGene00021681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>108769</td>\n",
       "      <td>109264</td>\n",
       "      <td>-</td>\n",
       "      <td>F53G12.1.1</td>\n",
       "      <td>WBGene00004274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>133582</td>\n",
       "      <td>134271</td>\n",
       "      <td>+</td>\n",
       "      <td>F53G12.5b.1</td>\n",
       "      <td>WBGene00003229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>14823733</td>\n",
       "      <td>14824009</td>\n",
       "      <td>+</td>\n",
       "      <td>Y54E5B.4.1</td>\n",
       "      <td>WBGene00006711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>14833448</td>\n",
       "      <td>14834697</td>\n",
       "      <td>-</td>\n",
       "      <td>F32A7.3a.1</td>\n",
       "      <td>WBGene00009304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>14951567</td>\n",
       "      <td>14952030</td>\n",
       "      <td>+</td>\n",
       "      <td>ZK909.2h.1</td>\n",
       "      <td>WBGene00002189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>14965021</td>\n",
       "      <td>14965417</td>\n",
       "      <td>+</td>\n",
       "      <td>ZK909.4.1</td>\n",
       "      <td>WBGene00000469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>15001585</td>\n",
       "      <td>15001631</td>\n",
       "      <td>+</td>\n",
       "      <td>ZK337.1b.1</td>\n",
       "      <td>WBGene00013969</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>507 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        start       end strand         name            gene\n",
       "0       22939     24650      -  Y74C9A.4a.1  WBGene00022278\n",
       "1       80344     81000      +  Y48G1C.2b.1  WBGene00000812\n",
       "2      101245    101504      -  Y48G1C.8a.1  WBGene00021681\n",
       "3      108769    109264      -   F53G12.1.1  WBGene00004274\n",
       "4      133582    134271      +  F53G12.5b.1  WBGene00003229\n",
       "..        ...       ...    ...          ...             ...\n",
       "502  14823733  14824009      +   Y54E5B.4.1  WBGene00006711\n",
       "503  14833448  14834697      -   F32A7.3a.1  WBGene00009304\n",
       "504  14951567  14952030      +   ZK909.2h.1  WBGene00002189\n",
       "505  14965021  14965417      +    ZK909.4.1  WBGene00000469\n",
       "506  15001585  15001631      +   ZK337.1b.1  WBGene00013969\n",
       "\n",
       "[507 rows x 5 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdf = pd.DataFrame([vars(t) for t in test], columns=[\"start\", \"end\", \"strand\", \"attributes\"])\n",
    "tdf[\"gene\"] = tdf[\"attributes\"].apply(lambda x: x.get(\"ID\")[0])\n",
    "tdf = pd.merge(tdf, gdf, left_on=\"gene\", right_on=\"name\")\n",
    "tdf = tdf.rename({'gene_y': \"gene\"}, axis=1).drop([\"gene_x\", \"attributes\"], axis=1).set_index(\"gene\")\n",
    "tdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "744bf8c9",
   "metadata": {},
   "source": [
    "Aggregate transcripts for minimum start base and maximum end base to capture full extent of the UTR for each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "45c174e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000010</th>\n",
       "      <td>-</td>\n",
       "      <td>11388803</td>\n",
       "      <td>11389140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000084</th>\n",
       "      <td>+</td>\n",
       "      <td>7608007</td>\n",
       "      <td>7609295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000148</th>\n",
       "      <td>+</td>\n",
       "      <td>10345475</td>\n",
       "      <td>10345594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000150</th>\n",
       "      <td>-</td>\n",
       "      <td>5342174</td>\n",
       "      <td>5342480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000156</th>\n",
       "      <td>+</td>\n",
       "      <td>8055142</td>\n",
       "      <td>8055482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00219288</th>\n",
       "      <td>-</td>\n",
       "      <td>13526268</td>\n",
       "      <td>13527384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00219313</th>\n",
       "      <td>+</td>\n",
       "      <td>9416506</td>\n",
       "      <td>9416572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00235158</th>\n",
       "      <td>-</td>\n",
       "      <td>2148830</td>\n",
       "      <td>2150194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00302974</th>\n",
       "      <td>+</td>\n",
       "      <td>532853</td>\n",
       "      <td>533212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00305050</th>\n",
       "      <td>-</td>\n",
       "      <td>1379101</td>\n",
       "      <td>1379231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>496 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end\n",
       "gene                                     \n",
       "WBGene00000010      -  11388803  11389140\n",
       "WBGene00000084      +   7608007   7609295\n",
       "WBGene00000148      +  10345475  10345594\n",
       "WBGene00000150      -   5342174   5342480\n",
       "WBGene00000156      +   8055142   8055482\n",
       "...               ...       ...       ...\n",
       "WBGene00219288      -  13526268  13527384\n",
       "WBGene00219313      +   9416506   9416572\n",
       "WBGene00235158      -   2148830   2150194\n",
       "WBGene00302974      +    532853    533212\n",
       "WBGene00305050      -   1379101   1379231\n",
       "\n",
       "[496 rows x 3 columns]"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_utrs = tdf.groupby([\"gene\", \"strand\"]).agg({\"start\": min, \"end\": max}).reset_index(\"strand\")\n",
    "\n",
    "# Forward strand UTR start bases match those of canonical UTRs minus 1\n",
    "test_utrs.loc[test_utrs[\"strand\"]==\"+\", \"start\"] += 1\n",
    "test_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5c1696",
   "metadata": {},
   "source": [
    "## Prepare control dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3279699",
   "metadata": {},
   "source": [
    "Create a dataframe of canonical WormBase control UTRs with a \"name\" column consisting of the parent gene's name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "d2d77389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4116</td>\n",
       "      <td>4220</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.3.1</td>\n",
       "      <td>WBGene00022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16586</td>\n",
       "      <td>16833</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.2</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16586</td>\n",
       "      <td>16837</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16702</td>\n",
       "      <td>16793</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.3</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17484</td>\n",
       "      <td>17910</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.4a.1</td>\n",
       "      <td>WBGene00022278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4192</th>\n",
       "      <td>15040474</td>\n",
       "      <td>15040608</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.1.1</td>\n",
       "      <td>WBGene00000881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>15041433</td>\n",
       "      <td>15041884</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.3.1</td>\n",
       "      <td>WBGene00009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>15050797</td>\n",
       "      <td>15051145</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.4.1</td>\n",
       "      <td>WBGene00009286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4195</th>\n",
       "      <td>15053608</td>\n",
       "      <td>15053793</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.6a.1</td>\n",
       "      <td>WBGene00009288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4196</th>\n",
       "      <td>15053619</td>\n",
       "      <td>15053672</td>\n",
       "      <td>+</td>\n",
       "      <td>F31C3.5.1</td>\n",
       "      <td>WBGene00009287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4197 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         start       end strand         name            gene\n",
       "0         4116      4220      -   Y74C9A.3.1  WBGene00022277\n",
       "1        16586     16833      +  Y74C9A.2a.2  WBGene00022276\n",
       "2        16586     16837      +  Y74C9A.2a.1  WBGene00022276\n",
       "3        16702     16793      +  Y74C9A.2a.3  WBGene00022276\n",
       "4        17484     17910      -  Y74C9A.4a.1  WBGene00022278\n",
       "...        ...       ...    ...          ...             ...\n",
       "4192  15040474  15040608      -    F31C3.1.1  WBGene00000881\n",
       "4193  15041433  15041884      -    F31C3.3.1  WBGene00009285\n",
       "4194  15050797  15051145      -    F31C3.4.1  WBGene00009286\n",
       "4195  15053608  15053793      -   F31C3.6a.1  WBGene00009288\n",
       "4196  15053619  15053672      +    F31C3.5.1  WBGene00009287\n",
       "\n",
       "[4197 rows x 5 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdf = pd.DataFrame([vars(c) for c in control], columns=[\"start\", \"end\", \"source\", \"strand\", \"attributes\"])\n",
    "cdf = cdf[cdf.source == \"WormBase\"]\n",
    "cdf['name'] = cdf[\"attributes\"].apply(lambda x: x.get(\"Parent\")[0].split(\":\")[-1])\n",
    "cdf = cdf.drop(['attributes', \"source\"], axis=1)\n",
    "cdf = pd.merge(cdf, gdf, on=\"name\")\n",
    "cdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfb881ad",
   "metadata": {},
   "source": [
    "Aggregate transcripts for minimum start base and maximum end base to capture full extent of the UTR for each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "c1995505",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000001</th>\n",
       "      <td>+</td>\n",
       "      <td>5109948</td>\n",
       "      <td>5110183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000006</th>\n",
       "      <td>-</td>\n",
       "      <td>2571740</td>\n",
       "      <td>2571987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000010</th>\n",
       "      <td>-</td>\n",
       "      <td>11388682</td>\n",
       "      <td>11389140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000012</th>\n",
       "      <td>-</td>\n",
       "      <td>3886070</td>\n",
       "      <td>3886110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000020</th>\n",
       "      <td>-</td>\n",
       "      <td>2255921</td>\n",
       "      <td>2256151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00304996</th>\n",
       "      <td>+</td>\n",
       "      <td>5341995</td>\n",
       "      <td>5342045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00305018</th>\n",
       "      <td>-</td>\n",
       "      <td>9588662</td>\n",
       "      <td>9588719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306080</th>\n",
       "      <td>-</td>\n",
       "      <td>2939477</td>\n",
       "      <td>2939518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306081</th>\n",
       "      <td>+</td>\n",
       "      <td>6872119</td>\n",
       "      <td>6872142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306121</th>\n",
       "      <td>-</td>\n",
       "      <td>231578</td>\n",
       "      <td>231691</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2593 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end\n",
       "gene                                     \n",
       "WBGene00000001      +   5109948   5110183\n",
       "WBGene00000006      -   2571740   2571987\n",
       "WBGene00000010      -  11388682  11389140\n",
       "WBGene00000012      -   3886070   3886110\n",
       "WBGene00000020      -   2255921   2256151\n",
       "...               ...       ...       ...\n",
       "WBGene00304996      +   5341995   5342045\n",
       "WBGene00305018      -   9588662   9588719\n",
       "WBGene00306080      -   2939477   2939518\n",
       "WBGene00306081      +   6872119   6872142\n",
       "WBGene00306121      -    231578    231691\n",
       "\n",
       "[2593 rows x 3 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_utrs = cdf.groupby([\"gene\", \"strand\"]).agg({\"start\": min, \"end\": max}).reset_index(\"strand\")\n",
    "control_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f36fb4d7",
   "metadata": {},
   "source": [
    "# Compare the two datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "078224b6",
   "metadata": {},
   "source": [
    "Now that we have two comparable datasets, we can calculate some statistical differences. We determine length of each UTR in both dataframes, and determine differences of test UTR length to true UTR length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "52ebdb54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>len</th>\n",
       "      <th>difference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000010</th>\n",
       "      <td>-</td>\n",
       "      <td>11388803</td>\n",
       "      <td>11389140</td>\n",
       "      <td>337</td>\n",
       "      <td>-121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000084</th>\n",
       "      <td>+</td>\n",
       "      <td>7608007</td>\n",
       "      <td>7609295</td>\n",
       "      <td>1288</td>\n",
       "      <td>794.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000148</th>\n",
       "      <td>+</td>\n",
       "      <td>10345475</td>\n",
       "      <td>10345594</td>\n",
       "      <td>119</td>\n",
       "      <td>-60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000150</th>\n",
       "      <td>-</td>\n",
       "      <td>5342174</td>\n",
       "      <td>5342480</td>\n",
       "      <td>306</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000156</th>\n",
       "      <td>+</td>\n",
       "      <td>8055142</td>\n",
       "      <td>8055482</td>\n",
       "      <td>340</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00219288</th>\n",
       "      <td>-</td>\n",
       "      <td>13526268</td>\n",
       "      <td>13527384</td>\n",
       "      <td>1116</td>\n",
       "      <td>1098.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00219313</th>\n",
       "      <td>+</td>\n",
       "      <td>9416506</td>\n",
       "      <td>9416572</td>\n",
       "      <td>66</td>\n",
       "      <td>-131.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00235158</th>\n",
       "      <td>-</td>\n",
       "      <td>2148830</td>\n",
       "      <td>2150194</td>\n",
       "      <td>1364</td>\n",
       "      <td>995.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00302974</th>\n",
       "      <td>+</td>\n",
       "      <td>532853</td>\n",
       "      <td>533212</td>\n",
       "      <td>359</td>\n",
       "      <td>128.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00305050</th>\n",
       "      <td>-</td>\n",
       "      <td>1379101</td>\n",
       "      <td>1379231</td>\n",
       "      <td>130</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>496 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end   len  difference\n",
       "gene                                                       \n",
       "WBGene00000010      -  11388803  11389140   337      -121.0\n",
       "WBGene00000084      +   7608007   7609295  1288       794.0\n",
       "WBGene00000148      +  10345475  10345594   119       -60.0\n",
       "WBGene00000150      -   5342174   5342480   306         6.0\n",
       "WBGene00000156      +   8055142   8055482   340         7.0\n",
       "...               ...       ...       ...   ...         ...\n",
       "WBGene00219288      -  13526268  13527384  1116      1098.0\n",
       "WBGene00219313      +   9416506   9416572    66      -131.0\n",
       "WBGene00235158      -   2148830   2150194  1364       995.0\n",
       "WBGene00302974      +    532853    533212   359       128.0\n",
       "WBGene00305050      -   1379101   1379231   130         NaN\n",
       "\n",
       "[496 rows x 5 columns]"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_utrs[\"len\"] = abs(control_utrs[\"end\"] - control_utrs[\"start\"])\n",
    "test_utrs[\"len\"] = abs(test_utrs[\"end\"] - test_utrs[\"start\"])\n",
    "test_utrs[\"difference\"] = (test_utrs[\"len\"] - control_utrs[\"len\"])\n",
    "test_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec2508a",
   "metadata": {},
   "source": [
    "## Supplementary Table 2 statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "c2e294c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical utrs that are missed by UTRme: 2119 (81.72%)\n",
      "UTRme utrs that match true utr to within 50 bases: 157 (6.05%)\n",
      "UTRme utrs that reduce true utr by >50 bases: 121 (4.67%)\n",
      "UTRme utrs that extend true utr by between 50 and 200 bases: 19 (0.73%)\n",
      "UTRme utrs that extend true utr by >200 bases: 177 (6.83%)\n",
      "UTRme utrs called where no true utr existed previously: 22\n",
      "Total UTRme UTRs: 496\n"
     ]
    }
   ],
   "source": [
    "total_test = len(test_utrs)\n",
    "total_control = len(control_utrs)\n",
    "\n",
    "genes_missed_utr = control_utrs.index.difference(test_utrs.index)\n",
    "num_missed_utrs = len(genes_missed_utr)\n",
    "genes_no_utr = set(gdf[\"gene\"]).difference(control_utrs.index)\n",
    "novel_utr_genes = genes_no_utr.intersection(test_utrs.index)\n",
    "num_novel_utrs = len(novel_utr_genes)\n",
    "\n",
    "BP1 = 50\n",
    "BP2 = 200\n",
    "num_matched =  len(test_utrs[abs(test_utrs[\"difference\"]) <= BP1])\n",
    "num_reduced = len(test_utrs[test_utrs['difference'] < -BP1])\n",
    "num_extended1 = len(test_utrs[(test_utrs['difference'] > BP1) & (test_utrs['difference'] <= BP2)])\n",
    "num_extended2 = len(test_utrs[test_utrs['difference'] > BP2])\n",
    "\n",
    "print(\"Canonical utrs that are missed by UTRme: {} ({:.2f}%)\".format(num_missed_utrs, 100*num_missed_utrs/total_control))\n",
    "print(\"UTRme utrs that match true utr to within {} bases: {} ({:.2f}%)\".format(BP1, num_matched, 100*num_matched/total_control))\n",
    "print(\"UTRme utrs that reduce true utr by >{} bases: {} ({:.2f}%)\".format(BP1, num_reduced, 100*num_reduced/total_control))\n",
    "print(\"UTRme utrs that extend true utr by between {} and {} bases: {} ({:.2f}%)\".format(BP1, BP2, num_extended1, 100*num_extended1/total_control))\n",
    "print(\"UTRme utrs that extend true utr by >{} bases: {} ({:.2f}%)\".format(BP2, num_extended2, 100*num_extended2/total_control))\n",
    "print(\"UTRme utrs called where no true utr existed previously: {}\".format(num_novel_utrs))\n",
    "print(\"Total UTRme UTRs: {}\".format(total_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9a5e30e",
   "metadata": {},
   "source": [
    "Save list of gene ids with novel 3' UTRs called by UTRme"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "68e9ae72",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WBGene00000632\n",
      "WBGene00001745\n",
      "WBGene00003977\n",
      "WBGene00006044\n",
      "WBGene00007976\n",
      "WBGene00010291\n",
      "WBGene00011651\n",
      "WBGene00011934\n",
      "WBGene00012389\n",
      "WBGene00012483\n",
      "WBGene00013404\n",
      "WBGene00016727\n",
      "WBGene00017908\n",
      "WBGene00018596\n",
      "WBGene00019992\n",
      "WBGene00020556\n",
      "WBGene00021473\n",
      "WBGene00022112\n",
      "WBGene00022367\n",
      "WBGene00044348\n",
      "WBGene00044435\n",
      "WBGene00305050\n"
     ]
    }
   ],
   "source": [
    "with open(os.path.join(out_dir, \"novel_utr_genes_utrme.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(novel_utr_genes)):\n",
    "        print(g)\n",
    "        f.write(g + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
